src/server.c

```diff
@@ -680,6 +680,50 @@ int clientsCronResizeQueryBuffer(client *c) {
     return 0;
 }
 
+/* The client output buffer can be adjusted to better fit the memory requierments.
+ *
+ * the logic is:
+ * in case the last observed peak size of the buffer equals the buffer size - we double the size
+ * in case the last observed peak size of the buffer is less than half the buffer size - we shrink by half.
+ * in any case we will reset the peak value each cron run.
+ * The function always returns 0 as it never terminates the client. */
+int clientsCronResizeOutputBuffer(client *c) {
+
+    void *new_buf = NULL;
+    const size_t BUFFER_TARGET_SHRINK_SIZE = c->buf_usable_size/2;
+    const size_t BUFFER_TARGET_EXPEND_SIZE = c->buf_usable_size*2;
+
+    if(c->bufpos != 0)  return 0;
```

Comment:
i think this check could mean that in some cases, we'll never manage to shrink the buffer in some cases, since there's a chance it'll always have data.

this brings me to two other realizations:
1. when bufpos is 0, there's no need for a realloc (it might be faster to free and malloc (no need to memmove).
2. i see we update buf_peak in _writeToClient. this means that when we append to the buffer, it's used size (`bufpos`) can be larger than the peak. so when we remove this check (`bugpos != 0`), we better move the updates of the peak to `_addReplyToBuffer`

Reply:
I also don't understand this check. `bufpos` isn't relevant to us. Only the peak is relevant. We might want to add an assert that `bufpos` <=  `buf_peak`.
We might also want to add an assert that `buf_preak` <= `buf_usable_size`.

Regarding `realloc` I think it'll be always faster to `zmalloc`, `memcpy` `bufpos` bytes and then `free` than to use `realloc`.

Reply:
@oranagra I agree. this was initially panic check and I agree it can be removed.
regarding moving the peak handling into _addReplyToBuffer I will do that.

Reply:
@yoav-steinberg i suggested to use free+malloc, in case `bufpos` is 0, instead of recalloc.
not to use malloc+memcpy+free.

Reply:
I have implemented the suggestion. If that looks reasonable to you I will resolve that case issue.